### PR TITLE
make findmin/findmax behavior match min/max (fix #23209)

### LINKED ISSUE #23209: make findmin/findmax behavior match min/max
See https://github.com/JuliaLang/julia/pull/23155. These should be made to match and documentation updated accordingly.

### PR DESCRIPTION
**Pull request by KlausC**
url: #23209

Same issue for sparse matrices not yet included - will follow.

### CODE REVIEW

File: base/array.jl [hunk 1]
```diff
@@ -1721,10 +1722,10 @@ function findmax(a)
     mi = i = 1
     m, s = next(a, s)
     while !done(a, s)
+        m != m && break
         ai, s = next(a, s)
         i += 1
-        ai != ai && continue # assume x != x => x is a NaN
-        if m != m || isless(m, ai)
+        if ai != ai || isless(m, ai)
```
Comment (StefanKarpinski) [inline base/array.jl]: In the `ai != ai` case this can short-circuit, right?

Comment (KlausC) [inline base/array.jl]: if `ai != ai` I expect `isless` is not called. That shall cover the `NaN`- case. This can occur at most once, since the next loop circle is broken by the `m != m` condition few lines before.

Comment (StefanKarpinski) [inline base/array.jl]: Ah, yes, I didn't notice the `m != m && break` check previously.

File: base/reducedim.jl [hunk 2]
```diff
@@ -57,24 +57,33 @@ function reduced_indices0(inds::Indices{N}, region) where N
...
 end
 
-
 ###### Generic reduction functions #####
 
+## extensions to the standard functions for BigInt, String and others
+
+zero2(t::Type) =  zero(t)
+zero2(::Type{T}) where T<:Number = zero(T)
+zero2(::Type{T}) where T<:AbstractString = ""
```
Comment (tkelman) [inline base/reducedim.jl]: requiring this is a bit of a red flag, is there any way to avoid needing this?

Comment (StefanKarpinski) [inline base/reducedim.jl]: The way is not to rely on zero behavior at all, but to use the first slice as the start value.

Comment (KlausC) [inline base/reducedim.jl]: let me improve this!

Comment (KlausC) [inline base/reducedim.jl]: The definition of `zero2` will be removed in the next commit, it is not required any longer.
I guess the red flag is risen, because a bug fix shall not introduce new `Base` features?

Comment (StefanKarpinski) [inline base/reducedim.jl]: As long as this isn't exported, it's not a feature addition. It just seems fishy to need a function like zero but for which `""` is the answer for strings. (Nevermind that `""` is a reasonable 1 value for strings; there is no zero value unless you embed strings into regular patterns, in which case the empty pattern is the zero value – but it's not a string.)

**Comment by KlausC** [comment]
Tests failed because of buggy `_findz` in `sparse/sparsematrix.jl:1831`. Will investigate with `@inbounds`removed.

File: test/distributed.jl [hunk 3]
```diff
@@ -3,6 +3,10 @@
 # Run the distributed test outside of the main driver since it needs its own
 # set of dedicated workers.
 
+using Base.Test
+
+@test open("distributed_exec.jl") do io; true end
```
Comment (tkelman) [inline test/distributed.jl]: what happened here, how is this related?

**Comment by tkelman** [comment]
The "No tests" is a consequence of #20362, not related. The failure was actually in offsetarray, but we seem to be having a problem in our testsystem right now where it isn't showing the actual backtrace of failures. It might work better if you run locally and only do `make test-offsetarray`, when it's running only a single test file the execution is a little less complicated.

**Comment by KlausC** [comment]
Fix complete. Please review.

File: test/reducedim.jl [hunk 4]
```diff
@@ -117,22 +117,201 @@ R = reducedim((a,b) -> a+b, [1 2; 3 4], 2, 0.0)
...
+    @test string(minimum!(copy(rval), A, init=false)) == string(rval)
 end
 
-for (tup, rval, rind) in [((1,), [5.0 3.0 6.0], [2 3 5]),
+for (tup, rval, rind) in [((1,), [5.0 5.0 6.0], [2 3 5]),
                           ((2,), reshape([6.0,5.0], 2, 1), reshape([5,2], 2, 1)),
                           ((1,2), fill(6.0,1,1),fill(5,1,1))]
     @test findmax(A, tup) == (rval, rind)
     @test findmax!(similar(rval), similar(rind), A) == (rval, rind)
+    @test string(maximum(A, tup)) == string(rval)
```
Comment (tkelman) [inline test/reducedim.jl]: why are you comparing all of these as strings?

Comment (KlausC) [inline test/reducedim.jl]: The comparisons using `==` do not work, when the result collections contain `NaN`, because `NaN != NaN`. Comparison with `===` does not compare array elements. 
And I wanted to be able to identify the failed tests by the test values. With a specialized compare function that was not the case ( `@test eq1(maximum(A, tup), rval)` prints its line literally).

Comment (StefanKarpinski) [inline test/reducedim.jl]: Can't you use `isequal` for these cases?

Comment (KlausC) [inline test/reducedim.jl]: Yes, I did it. Actually it revealed some differences between `Array{T,2}` 2x1 and `Array{T,1}` with equal contents.

**Comment by KlausC** [comment]
Are there more changes to do for this fix?

**Comment by mbauman** [comment]
Very nice work!  This is definitely needed.  I took the liberty of removing the unrelated change in tests/distributed.jl.  My only question is if we want to deprecate this behavior first.  It's a little tricky (especially in the case of all-nan elements), but I think it should be doable.

**Comment by KlausC** [comment]
Is there anything, I should or could do about that question?

**Comment by StefanKarpinski** [comment]
Your call @mbauman – I'm not sure if it's worth it, but if you think so, go for it.

**Comment by mbauman** [comment]
On balance, I think we should just leave this as a breaking change.  I've added a NEWS.md bullet point.

**Comment by StefanKarpinski** [comment]
Thank you so much @KlausC for seeing this through!
